src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        bytesPerRow: 256,
+        rowsPerImage: 16,
+        dataSize: 1,
+        method: method,
+      });
+    }, !_success);
+  });
+
+g.test('mip_level_must_be_in_range')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('mipLevelCount', [3, 5]))
+      .combine(poptions('mipLevel', [3, 4]))
+  )
+  .fn(async t => {
+    const { mipLevelCount, mipLevel, method } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 32, height: 32, depth: 1 },
+      mipLevelCount: mipLevelCount,
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    const _success = mipLevel < mipLevelCount;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        mipLevel: mipLevel,
+        bytesPerRow: 256,
+        rowsPerImage: 16,
+        dataSize: 1,
+        method: method,
+      });
+    }, !_success);
+  });
+
+g.test('1d_texture')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('copyWidth', [0, 1]))
+      .combine(poptions('copyHeight', [0, 1]))
+      .combine(poptions('copyDepth', [0, 1]))
+  )
+  .fn(async t => {
+    const { copyWidth, copyHeight, copyDepth, method } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 2, height: 1, depth: 1 },
+      mipLevelCount: 1,
+      dimension: '1d',
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    // For 1d textures we require copyHeight and copyDepth to be 1,
+    // copyHeight or copyDepth being 0 should cause a validation error.
+    const _success = copyHeight == 1 && copyDepth == 1;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: copyWidth, height: copyHeight, depth: copyDepth },
+        mipLevel: 0,
+        bytesPerRow: 256,
+        rowsPerImage: 16,
+        dataSize: 16,
+        method: method,
+      });
+    }, !_success);
+  });
+
+enum Coord {
+  x = 'x',
+  y = 'y',
+  z = 'z',
+}
+
+const kCoords = [Coord.x, Coord.y, Coord.z];
+
+const kTestValuesForDivisibilityBy4 = [1, 2, 3, 4, 6, 8, 12];
+
+g.test('texel_block_alignemnts_on_origin')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('value', kTestValuesForDivisibilityBy4))
+      .combine(poptions('coordinate_to_test', kCoords))
```

Comment:
Not sure the enum is needed, could maybe just be:
```suggestion
      .combine(poptions('coordinate_to_test', ['x', 'y', 'z'] as const))
```

Reply:
Done